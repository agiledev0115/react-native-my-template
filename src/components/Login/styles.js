import { StyleSheet } from 'react-native';
import Colors from '../../helpers/Colors';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: Colors.primary,
  },
  formContainer: {
    alignSelf: 'stretch',
    justifyContent: 'center',
    alignItems: 'flex-start',
    backgroundColor: Colors.white,
    marginHorizontal: 40,
    padding: 20,
  },
});

export default styles;
