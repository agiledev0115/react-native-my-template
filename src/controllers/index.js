export { HttpClient } from '@/controllers/HttpClient';
export { UserController } from '@/controllers/UserController';
