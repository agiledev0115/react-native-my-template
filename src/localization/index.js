export { strings } from '@/localization/Localization';
