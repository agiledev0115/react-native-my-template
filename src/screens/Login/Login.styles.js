import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    paddingHorizontal: 40,
  },
  formContainer: {
    borderRadius: 5,
    padding: 20,
    width: '100%',
  },
  submitButton: {
    marginTop: 20,
  },
});
