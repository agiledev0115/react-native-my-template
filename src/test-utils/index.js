export { configureStore } from '@/test-utils/configureStore';
export { withProviders } from '@/test-utils/render';
