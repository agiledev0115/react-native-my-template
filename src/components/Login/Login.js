import React, { Component } from 'react';
import {
  View,
  Text,
} from 'react-native';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import App from '../../App';
import Button from '../common/Button';
import TextField from '../common/TextField';
import ErrorView from '../common/ErrorView';
import ShadowStyles from '../../helpers/ShadowStyles';
import TextStyles from '../../helpers/TextStyles';
import strings from '../../localization';
import { login, actionTypes } from '../../actions/UserActions';
import getUser from '../../selectors/UserSelectors';
import loadingSelector from '../../selectors/LoadingSelector';
import { errorsSelector } from '../../selectors/ErrorSelector';
import styles from './styles';

class Login extends Component {
  static navigatorStyle = {
    navBarHidden: true,
  };

  static getDerivedStateFromProps(nextProps) {
    if (nextProps.user !== null) {
      App.startLoggedInApp();
    }
    return null;
  }

  constructor() {
    super();
    this.state = {
      email: '',
      password: '',
    };
  }

  passwordChanged = value => this.setState({ password: value });

  emailChanged = value => this.setState({ email: value });

  login = () => this.props.login(this.state.email, this.state.password);

  render() {
    const { isLoading, errors } = this.props;
    return (
      <View style={styles.container}>
        <View style={[styles.formContainer, ShadowStyles.shadow]}>
          <Text style={TextStyles.fieldTitle}>
            {strings.email}
          </Text>
          <TextField
            placeholder={strings.email}
            onChangeText={this.emailChanged}
            value={this.state.email}
          />
          <Text style={TextStyles.fieldTitle}>
            {strings.password}
          </Text>
          <TextField
            placeholder={strings.password}
            value={this.state.password}
            onChangeText={this.passwordChanged}
            secureTextEntry
          />
          <ErrorView errors={errors} />
          <Button
            onPress={this.login}
            title={isLoading ? strings.loading : strings.login}
          />
        </View>
      </View>
    );
  }
}

Login.propTypes = {
  login: PropTypes.func.isRequired,
  user: PropTypes.object,
  isLoading: PropTypes.bool.isRequired,
  errors: PropTypes.array,
};

Login.defaultProps = {
  user: null,
  errors: [],
};

const mapStateToProps = state => ({
  user: getUser(state),
  isLoading: loadingSelector([actionTypes.LOGIN])(state),
  errors: errorsSelector([actionTypes.LOGIN])(state),
});

const mapDispatchToProps = dispatch => ({
  login: (email, password) => dispatch(login(email, password)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Login);
