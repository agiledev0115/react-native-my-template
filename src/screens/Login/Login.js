import { useTheme } from '@react-navigation/native';
import React, { useState } from 'react';
import { View } from 'react-native';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import { login, TYPES } from '@/actions/UserActions';
import { Button, ErrorView, TextField } from '@/components';
import { strings } from '@/localization';
import { styles } from '@/screens/Login/Login.styles';
import { errorsSelector } from '@/selectors/ErrorSelectors';
import { isLoadingSelector } from '@/selectors/StatusSelectors';
import { shadow } from '@/theme';

export function Login() {
  const { colors } = useTheme();
  const dispatch = useDispatch();
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const errors = useSelector((state) => errorsSelector([TYPES.LOGIN], state), shallowEqual);
  const isLoading = useSelector((state) => isLoadingSelector([TYPES.LOGIN], state));

  const handleSubmit = () => {
    dispatch(login(username, password));
  };

  return (
    <View style={styles.container}>
      <View style={[styles.formContainer, shadow.primary, { backgroundColor: colors.primary }]}>
        <TextField
          autoCapitalize="none"
          accessibilityHint={strings.login.usernameHint}
          accessibilityLabel={strings.login.username}
          onChangeText={setUsername}
          placeholder={strings.login.username}
          value={username}
        />
        <TextField
          secureTextEntry
          accessibilityHint={strings.login.passwordHint}
          accessibilityLabel={strings.login.password}
          autoCapitalize="none"
          onChangeText={setPassword}
          placeholder={strings.login.password}
          textContentType="password"
          value={password}
        />
        <ErrorView errors={errors} />
        <Button
          onPress={handleSubmit}
          style={styles.submitButton}
          title={isLoading ? strings.common.loading : strings.login.button}
        />
      </View>
    </View>
  );
}
