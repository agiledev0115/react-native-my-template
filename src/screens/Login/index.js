import Login from '_screens/Login/Login';

export default Login;
