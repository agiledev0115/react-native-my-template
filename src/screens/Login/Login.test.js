import { fireEvent, render, waitFor } from '@testing-library/react-native';
import React from 'react';
import { UserController } from '@/controllers';
import { strings } from '@/localization';
import { Login } from '@/screens/Login/Login';
import { withProviders } from '@/test-utils';

jest.mock('@/controllers/UserController');

describe('Login', () => {
  const fakeData = {
    username: 'username',
    password: 'password',
  };

  it('should match the snapshot', () => {
    const { toJSON } = render(withProviders(<Login />));

    expect(toJSON()).toMatchSnapshot();
  });

  it('should submit correctly with valid username and password', async () => {
    const loginSpy = jest.spyOn(UserController, 'login');

    const { getByHintText, getByText } = render(withProviders(<Login />));

    const submitButton = getByText(strings.login.button);
    const usernameInput = getByHintText(strings.login.usernameHint);
    const passwordInput = getByHintText(strings.login.passwordHint);

    fireEvent.changeText(usernameInput, fakeData.username);
    fireEvent.changeText(passwordInput, fakeData.password);

    await waitFor(() => {
      fireEvent.press(submitButton);
    });

    expect(loginSpy).toHaveBeenCalledWith(fakeData.username, fakeData.password);
  });

  it('should show error when password is not provided', async () => {
    const loginSpy = jest.spyOn(UserController, 'login');

    const { getByHintText, getByText } = render(withProviders(<Login />));

    const submitButton = getByText(strings.login.button);
    const usernameInput = getByHintText(strings.login.usernameHint);

    fireEvent.changeText(usernameInput, fakeData.username);

    await waitFor(() => {
      fireEvent.press(submitButton);
    });

    expect(loginSpy).toHaveBeenCalledWith(fakeData.username, '');
    expect(getByText(strings.login.invalidCredentials)).toBeTruthy();
  });
});
