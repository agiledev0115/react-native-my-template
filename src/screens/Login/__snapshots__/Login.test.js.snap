// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match the snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 40,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 5,
          "padding": 20,
          "width": "100%",
        },
        Object {
          "elevation": 5,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 3.84,
        },
        Object {
          "backgroundColor": "rgb(0, 122, 255)",
        },
      ]
    }
  >
    <TextInput
      accessibilityHint="Enter your username"
      accessibilityLabel="Username"
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="Username"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "color": "rgb(28, 28, 30)",
          },
          Object {
            "fontSize": 16,
            "fontWeight": "400",
          },
          Object {
            "borderBottomWidth": 0.5,
            "marginVertical": 15,
            "paddingVertical": 5,
          },
          null,
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      accessibilityHint="Enter your password"
      accessibilityLabel="Password"
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="Password"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Array [
          Object {
            "color": "rgb(28, 28, 30)",
          },
          Object {
            "fontSize": 16,
            "fontWeight": "400",
          },
          Object {
            "borderBottomWidth": 0.5,
            "marginVertical": 15,
            "paddingVertical": 5,
          },
          null,
        ]
      }
      textContentType="password"
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "rgb(216, 216, 216)",
          "borderRadius": 5,
          "borderWidth": 1,
          "justifyContent": "center",
          "marginTop": 20,
          "opacity": 1,
          "padding": 10,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "rgb(28, 28, 30)",
            },
            Object {
              "fontSize": 16,
              "fontWeight": "700",
            },
            null,
          ]
        }
      >
        Login
      </Text>
    </View>
  </View>
</View>
`;
